### ISSUE TITLE
Discuss population penalty for scoring function

### ISSUE THREAD

**Issue by davidlmobley** [open-forcefield-group/smarty#10]
As somewhat alluded to in #8 , we want to make some improvements (or what we believe are improvements - we want to test) to the scoring function. There's a strong feeling on this end that right now we're treating discovery of new 1:many matches as equivalent to new 1:1 matches, whereas the latter should be preferred in general, and that adjusting this treatment will lead to faster/better exploration.

We would like to try a couple of simple scoring function modifications in addition to #9 on normalization - specifically, we would like to try implementing a sub-optimal reward for introduction of a new pattern which matches multiple atom types, with an optimal reward only for a pattern which matches all of a given atom type and only that atom type.

We are initially thinking of a function which would give a reward of `x = f/(ln(n)+1)` where `n` is the number of atom types encompassed by the pattern, and `f` is the fraction of those atom types recovered (averaged across types). Thus:
- If a pattern matches only one type and all of that type, the reward is `1.0`
- If a pattern matches two types and all of those types, the reward is `1.0/(ln(2)+1)`
- If a pattern matches five types but only 50% of each of those types, the reward is `1.0/(ln(5)+1)`

The idea of `ln(n)` was because of a feeling that `1/n` falls off too rapidly for introduction of new sub-types that match multiple types.

**Comment by jchodera** [comment]
> There's a strong feeling on this end that right now we're treating discovery of new 1:many matches as equivalent to new 1:1 matches,

A `[#1]` atom type will match the largest class of hydrogen subtypes from parm@frosst. Addition of something for which there are a lot of matches---say, `$([#1])&$(*~[#6])`---will get a big score increase.  From there, the increases will be smaller because the matched atoms increase less and less.

I would, however, keep in mind that we are intending to replicate a scheme in which we are using experimental data in the likelihood function. That scenario doesn't really match any of the schemes you describe above, so I worry we are starting to get away from the Bayesian approach in this construction.

**Comment by davidlmobley** [comment]
> A [#1] atom type will match the largest class of hydrogen subtypes from parm@frosst. Addition of something for which there are a lot of matches---say, $([#1])&$(*~[#6])---will get a big score increase. From there, the increases will be smaller because the matched atoms increase less and less.

Yes. In terms of efficiency in discovering all of the relevant types this doesn't seem as efficient to us. Chris also has some concerns that this might result in an overemphasis of some regions of chemical space at the expense of others, though given your point below this is probably less of an issue: 

> I would, however, keep in mind that we are intending to replicate a scheme in which we are using experimental data in the likelihood function. That scenario doesn't really match any of the schemes you describe above, so I worry we are starting to get away from the Bayesian approach in this construction.

Can you explain what's more or less Bayesian about one particular approach to the scoring or objective function, here? I certainly understand the point about using experimental data in the likelihood function, and it's obviously key for where we're going, but I'm missing how it connects. 

Trying to connect the dots - I'm guessing what you're saying is that you're imagining that the atom types here are like one particular type of experimental data, and there is no reason you would want to weight data for one particular functional group differently from data for another functional group, so it would be the total amount of data available which would drive things. Is that right, @jchodera ?

**Comment by jchodera** [comment]
> Can you explain what's more or less Bayesian about one particular approach to the scoring or objective function, here?

The likelihood function is for parameters `\theta` given data `D = {D_1, D_2, ..., D_N}` is

```
L(\theta | D) = \prod_{n=1}^N p_n(D_n | \theta)
```

so that the log-likelihood is

```
logL(\theta | D) = \sum_{n=1}^N \log p_n(D_n | \theta)
```

The log-likelihood plays the role of the score in this SMARTY example. The score is _additive_ in the log probabilities of each observation. The current SMARTY score is equivalent to saying that you have N measurements---one per molecules---where the individual observation model probabilities `p_n(D_n | \theta)` that are close to 1 if the parameters/typing are good and `1 / (Natoms * e)` if the parameters are bad. The essential thing is that the log-likelihood is additive in the data.

**Comment by cbayly13** [comment]
Trying to follow the equations above. Here is my new suggestion:
Score for a smarts is 
sum(a_i)/(n+log(n)) 
where the sum goes over i=1,n ; n is the number of reference atom types matched by the smarts, a_i is the fractional occupancy of the refType count matched by the smarts with respect to the total refType count.

The only difference wrt a simple sum of fractional occupancy is the log(n) in the denominator, which gently punishes the smart for matching more than one atom type. The fractional occupancy normalizes the counts. 

Doubtless this needs further discussion but this is what I think now.

**Comment by cbayly13** [comment]
oops didn't mean to close

**Comment by davidlmobley** [comment]
Partly for my own reference:  I now have a concrete example now of one aspect of the issue Christopher was trying to raise here. His thought - and this made sense to me as well - is that, given the choice, we should pick SMARTS patterns which ONLY match the chemical environments they are being used to parameterize, and not other environments as well -- i.e., that there should be a 1:1 correspondence between SMARTS patterns and target atom types. This is not currently the case; a general SMARTS pattern which covers all the same ground as three atom types is scored just as well as specific SMARTS patterns which match each of those three types.

In any case, a concrete example of this comes up in AlkEthOH if I start it off with the SMARTS patterns which match each atom type uniquely, resulting in a 100% score. Here's where it starts off, in a T=0 run:

```
INDEX        ATOMS  MOLECULES                                                          TYPE NAME                           SMARTS
    1 :          0          0 |                                                         c_iodine                            [#53]
    2 :          0          0 |                                                        c_bromine                            [#35]
    3 :          0          0 |                                                       c_chlorine                            [#17]
    4 :          0          0 |                                                         c_sulfur                            [#16]
    5 :          0          0 |                                                    c_phosphorous                            [#15]
    6 :          0          0 |                                                       c_fluorine                             [#9]
    7 :          0          0 |                                                         c_oxygen                             [#8]
    8 :          0          0 |                                                       c_nitrogen                             [#7]
    9 :          0          0 |                                                         c_carbon                             [#6]
   10 :          0          0 |                                                       c_hydrogen                             [#1]
   11 :        244         38 |                                                c_hydrogen-carbon                      [$([#1]-C)]
   12 :        116         37 |                                            c_hydrogen-carbon-ewd  [$([#1]-C-[#7,#8,F,#16,Cl,Br])]
   13 :         33         22 |                                           c_hydrogen-carbon-ewd2 [$([#1]-C(-[#7,#8,F,#16,Cl,Br])-[#7,#8,F,#16,Cl,Br])]
   14 :          3          3 |                                           c_hydrogen-carbon-ewd3 [$([#1]-C(-[#7,#8,F,#16,Cl,Br])(-[#7,#8,F,#16,Cl,Br])-[#7,#8,F,#16,Cl,Br])]
   15 :         68         42 |                                                c_hydrogen-oxygen                    [#1$(*-[#8])]
   16 :        232         42 |                                                     c_carbon-tet                           [#6X4]
   17 :         39         30 |                                                   c_oxygen-dival                           [#8X2]
   18 :         68         42 |                                                c_oxygen-hydrogen                 [OX2+0$(*-[#1])]
TOTAL :        803         42
```

Naturally, it achieves 100% score by doing this, and, since this is T=0, the score isn't going to go down. Still, it can "forget" some "important" SMARTS patterns without degrading the score  -- it loses several of these by the end, including these:
- c_hydrogen-carbon                      [$([#1]-C)]
- c_carbon-tet                                 [#6X4]
- c_oxygen-dival                             [#8X2]

Of course, it loses these because the generic elemental types are there to pick up the slack so there is no loss in score (i.e., since there are only two kinds of specialized oxygen, you can delete one of the specialized oxygens and have the generic oxygen handle the other cases). The "problem" with this, from Christopher's perspective, is that:
a) the generics are there to handle what isn't handled by anything else, so having things be treated by a generic is less than ideal
b) now we're throwing out specific SMARTS patterns which DO have a 1:1 correspondence with atom types (chemical environments, if you will) in favor of general smarts patterns which match many atom types/chemical environments (but then some of these get overridden later)

Perhaps someone also might worry that this would make it harder to think about where individual parameters come from since instead of using a specialized carbon type which clearly indicates what it's parameterizing, you find out what carbons you're parameterizing with the generic type by looking at what carbons AREN'T covered by specialized types. **However**, maybe we've already burned that bridge when we committed to hierarchical parameters where the last one wins, as I think it inevitably leads us to this same place.

At first this seemed to me like a huge problem, but as I think about it more, it may not be so much, because:
a) if there is no data that says otherwise, then a generic carbon is as good as a specialized one, which is just what we're seeing here
b) the "last one wins" scheme we've already committed to means that we inevitably are going to have 1:many matches at some point and the key to success is getting enough of those overridden by specialized matches down the line that we find what we need to find
c) along the lines of (b), it's the **real data** (which we don't have here) which is going to provide the impetus to override generics with specialized types; here all we're after, at least initially, is coverage

That said, I wonder if we **do** need some sort of way to make sure our machinery has what it takes to develop specialized types when the data says so -- i.e., is there a way to drive the system to maintain (and create new) specialized types here when parm@frosst has them, to achieve closer to 1:1 correspondence?

**Comment by jchodera** [comment]
> His thought - and this made sense to me as well - is that, given the choice, we should pick SMARTS patterns which ONLY match the chemical environments they are being used to parameterize, and not other environments as well -- i.e., that there should be a 1:1 correspondence between SMARTS patterns and target atom types. This is not currently the case; a general SMARTS pattern which covers all the same ground as three atom types is scored just as well as specific SMARTS patterns which match each of those three types.

This might just be me being dense, but I have no idea what the problem is here. Each SMIRFF type is allowed to match up to one (and no more than one) reference type. The matching is done in a way that maximizes the number of matched atoms.

Can you try to explain this from a different angle? I am just not getting this somehow...

**Comment by davidlmobley** [comment]
> This might just be me being dense, but I have no idea what the problem is here. Each SMIRFF type is allowed to match up to one (and no more than one) reference type. The matching is done in a way that maximizes the number of matched atoms.
> 
> Can you try to explain this from a different angle? I am just not getting this somehow...

So take these two sets of possible SMIRFF types for hydrogen:
case 1:
- `[#1]` hydrogen 
- `[$([#1]-C)]` hydrogen-carbon
- `[#1$(*-[#8])]` hydrogen-oxygen

versus case 2:
- `[#1]` hydrogen
- `[#1$(*-[#8])]` hydrogen-oxygen

If you apply these to AlkEthOH with no other hydrogen types, both get equivalent scores, as the generic carbon covers everything that hydrogen-carbon and hydrogen-oxygen covers, so giving up hydrogen-carbon doesn't cost anything. But you've actually given up information about what chemical environment you are looking at by deleting "`[$([#1]-C)]` hydrogen-carbon". If you went through the atom types which COULD match the `[#1]` query you would find five of them, whereas if you went through and looked at atom types which COULD match the `[$([#1]-C)]` query you would find only four. Chris's view is that the latter, then, is _better_.

Why is it better? Partly, for efficiency of exploration. Specifically, it's going to be hard to discover the more sophisticated types `[$([#1]-C-[#7,#8,F,#16,Cl,Br])]` if you don't (a) have some incentive to retain `[$([#1]-C)]` when you discover it, and (b) have a way to decorate things you've already found (here, with electron withdrawing groups `[#7,#8,F,#16,Cl,Br]`). Right now, the data gives no incentive to retain `[$([#1]-C)]` whenever `[#1$(*-[#8])]` has already been found, as it provides **NO SCORE BENEFIT** to have `[$([#1]-C)]` since everything it matches is already covered by the generic `[#6]`.

Let's try it one more way too: 

Stop thinking about how you're scoring it for a moment. Imagine you went through all the SMARTS patterns you are currently using and, instead of doing what you do now (where you find the LAST SMARTS pattern which matches a particular reference type), you looked at ALL SMARTS patterns matching that reference type and counted how many SMARTS patterns in your set correspond to each individual reference type - i.e., how many hydrogen SMARTS strings you have would recognize this specific hydrogen?  Given the choice of two sets of SMARTS patterns which would match all reference types, wouldn't we prefer the set of SMARTS patterns where each SMARTS corresponds to the fewest reference types? In other words, if we had to choose just two types of hydrogen to match all AlkEthOH hydrogen types, we would want to match them with `[$([#1]-C)]` and `[#1$(*-[#8])]`, rather than `[#1]` and  `[#1$(*-[#8])]`, even though both in your current scoring scheme work exactly as well. Why? Most importantly, because it lends itself best to discovering new sophisticated types of carbon-attached hydrogen that you can't get to in a single step from `[#1]`, but also because it leads to SMARTS strings which provide more information about what environments they are supposed to be recognizing, rather than SMARTS strings which only say that they recognize the environment which isn't covered by any of the other environments.

If that doesn't help, we should talk by phone tomorrow if possible.

**Comment by davidlmobley** [comment]
John and I talked, and we are somewhat on the same page on this now, though he and Chris still need to be locked into a room with boxing gloves at some point to duke it out. Some take-aways:
- Because we're committed to hierarchical atom typing, there is no reason here that we can really prefer case 2 from above to case 1, and indeed in JDC's view case 1 is preferable since it has fewer atom types but still types everything in just the same way (for hierarchical reasons). 
- If there is no data that says otherwise, then a generic carbon is as good as a specialized one, which is just what we're seeing here
- there is no good reason to give generic (i.e. elemental) types special treatment and say that it's bad for them to be populated. CIB liked to use them as a "warning sign", but down the line, if we have generic types which aren't tightly constrained, parameter fluctuation for these itself will serve as a warning sign, and if they ARE tightly constrained, then they aren't a warning sign

So the problem is not one of scoring but data. And to large extent we already started down this road when we decided on hierarchical typing.
